src/pyhf/__init__.py

```diff
@@ -27,44 +27,77 @@ def get_backend():
 
 
 @events.register('change_backend')
-def set_backend(backend, custom_optimizer=None):
+def set_backend(backend, custom_optimizer=None, custom_backend=False, _session=None):
     """
     Set the backend and the associated optimizer
 
     Example:
         >>> import pyhf
         >>> import tensorflow as tf
-        >>> pyhf.set_backend(pyhf.tensor.tensorflow_backend(session=tf.Session()))
+        >>> sess = tf.compat.v1.Session()
+        >>> pyhf.set_backend("tensorflow", _session=sess)
 
     Args:
-        backend: One of the supported pyhf backends: NumPy, TensorFlow, and PyTorch
+        backend (`str` or `pyhf.tensor` backend): One of the supported pyhf backends: NumPy, TensorFlow, and PyTorch
+        custom_optimizer (`pyhf.optimize` optimizer): Optional custom optimizer defined by the user
+        custom_backend (`bool`): If :code:`True` the :code:`backend` argument is allowed to be a custom user defined backend outside of the supported backends
+        _session (|tf.compat.v1.Session|_): TensorFlow v1 compatible Session to use when the :code:`"tensorflow"` backend API is used
+
+    .. |tf.compat.v1.Session| replace:: ``tf.compat.v1.Session``
+    .. _tf.compat.v1.Session: https://www.tensorflow.org/api_docs/python/tf/compat/v1/Session
 
     Returns:
         None
     """
     global tensorlib
     global optimizer
 
+    supported_backend_types = (
+        tensor.numpy_backend,
+        tensor.tensorflow_backend,
+        tensor.pytorch_backend,
+    )
+    supported_backend_names = (backend().name for backend in supported_backend_types)
+
+    if isinstance(backend, str) and (
+        backend in supported_backend_names or backend == "numpy_minuit"
+    ):
+        if backend == "numpy":
+            backend = tensor.numpy_backend()
+        elif backend == "tensorflow":
+            backend = tensor.tensorflow_backend(session=_session)
+        elif backend == "pytorch":
+            backend = tensor.pytorch_backend()
+        elif backend == "numpy_minuit":
+            backend = tensor.numpy_backend(poisson_from_normal=True)
+    elif not isinstance(backend, supported_backend_types) and not custom_backend:
+        raise ValueError(
+            "'{}' is not a supported backend.\n             Select from one of the supported backends: numpy, tensorflow, pytorch".format(
+                backend
+            )
+        )
+
     # need to determine if the tensorlib changed or the optimizer changed for events
     tensorlib_changed = bool(backend.name != tensorlib.name)
     optimizer_changed = False
 
-    if backend.name == 'tensorflow':
+    # Check on instance of type instead of name to prevent custom name conflicts
```

Comment:
why not just check for `backend.name == "numpy"` or similar?

Reply:
If we are going to be allowing for the possibility of custom backends there is nothing keeping someone from giving any name. So if they also named their custom backend "numpy" this would improperly treat it as the supported numpy backend, no?